File: julia/Hand.jl
include("Card.jl")

# Represents a hand of playing cards.
# This class defines a hand object that can hold a collection of card objects.
# author: Davis Guest
mutable struct Hand
    cards    :: Vector{Card}
    sorted   :: Vector{Card}
    handType :: Int
end


# Initializes a new hand with an empty list of cards, its sorted variant, and a default hand type of 0.
function initHand()
    return Hand(Card[], Card[], 0)
end


# Returns a string representation of the hand.
function toString(self::Hand)
    str = ""
    for i = 1:5
        if (self.cards[i].rank != 10) ; str *= " " ; end

        str *= toString(self.cards[i])

        if (i == 1 || i % 5 != 0) ; str *= " " ; end
    end

    if (self.handType == 0)      ; return str
    elseif (self.handType == 10) ; str *= " - Royal Straight Flush"
    elseif (self.handType == 9)  ; str *= " - Straight Flush"
    elseif (self.handType == 8)  ; str *= " - Four of a Kind"
    elseif (self.handType == 7)  ; str *= " - Full House"
    elseif (self.handType == 6)  ; str *= " - Flush"
    elseif (self.handType == 5)  ; str *= " - Straight"
    elseif (self.handType == 4)  ; str *= " - Three of a Kind"
    elseif (self.handType == 3)  ; str *= " - Two Pair"
    elseif (self.handType == 2)  ; str *= " - Pair"
    else                         ; str *= " - High Card"
    end

    return str;
end

# Adds a card to the hand's list of cards.
function addCard(self::Hand, card::Card)
    push!(self.cards, card)
end


# Compares this hand with another hand based on their ranks.
# Used to sort each hand type and tiebreakers accordingly.
function compare(self::Hand, other::Hand)
    assessHand(self) , assessHand(other)

    typeComparison = self.handType - other.handType

    return compareHelper(self, other, typeComparison, 0)
end


# Recursive helper method for the compare_hand method.
function compareHelper(self::Hand, other::Hand, diff::Int, pass::Int)
    diff != 0 && return diff

    thisBreaker = getTieBreakerCard(self, pass)
    otherBreaker = getTieBreakerCard(other, pass)

    rankDiff = thisBreaker.rank - otherBreaker.rank
    suitDiff = thisBreaker.suit - otherBreaker.suit

    if (pass < 2 && (self.handType == 2 || self.handType == 3))
        rankDiff == 0 && return compareHelper(self, other, rankDiff, pass + 1)
        return rankDiff
    end

    rankDiff == 0 && return suitDiff
    return rankDiff;
end


# Analyzes the current collection of cards in the hand and determines its hand type.
# The sorted instance variable is updated to contain the cards sorted by rank.
# This method sets the handType instance variable.
function assessHand(self::Hand)
    self.sorted = sortHand(self.cards)

    if     isRoyalStraightFlush(self) ; self.handType = 10
    elseif isStraightFlush(self)      ; self.handType = 9
    elseif isFourOfAKind(self)        ; self.handType = 8
    elseif isFullHouse(self)          ; self.handType = 7
    elseif isFlush(self)              ; self.handType = 6
    elseif isStraight(self)           ; self.handType = 5
    elseif isThreeOfAKind(self)       ; self.handType = 4
    elseif isTwoPair(self)            ; self.handType = 3
    elseif isPair(self)               ; self.handType = 2
    else                              ; self.handType = 1
    end
end


# Determines if the hand is a royal straight flush.
function isRoyalStraightFlush(self::Hand)
    rankList = getRankList(self)

    return isStraight(self) && rankList[1] == 10 && rankList[5] == 14
end


# Determines if the hand is a straight flush.
function isStraightFlush(self::Hand)
    rankList = getRankList(self)

    return isStraight(self) && isFlush(self)
end


# Determines if the hand is a four of a kind.
function isFourOfAKind(self::Hand)
    rankList = getRankList(self)

    return (rankList[1] == rankList[4]) || (rankList[2] == rankList[5])
end


# Determines if the hand is a full house.
function isFullHouse(self::Hand)
    rankList = getRankList(self)

    return ((rankList[1] == rankList[2]) && (rankList[3] == rankList[5])) &&
           ((rankList[1] == rankList[3]) && (rankList[4] == rankList[5]))
end


# Determines if the hand is a flush.
function isFlush(self::Hand)
    rankList = getRankList(self)
    suit = self.sorted[1].suit

    for i = 1:5
        self.sorted[i].suit != suit && return false
    end 

    return self.sorted[5].suit == suit
end


# Determines if the hand is a straight.
function isStraight(self::Hand)
    rankList = getRankList(self)

    if (rankList[5] == 14 && rankList[1] == 2)
        rankList[5] = 1
        rankList = sort(rankList)
    end

    for i = 1:4
        (rankList[i + 1] != rankList[i] + 1) && return false
    end

    return true
end


# Determines if the hand is a three of a kind.
function isThreeOfAKind(self::Hand)
    rankList = getRankList(self)

    return (rankList[1] == rankList[3]) ||
           (rankList[2] == rankList[4]) ||
           (rankList[3] == rankList[5])
end


# Determines if the hand is a two pair.
function isTwoPair(self::Hand)
    rankList = getRankList(self)

    return ((rankList[1] == rankList[2]) && (rankList[3] == rankList[4])) ||
           ((rankList[1] == rankList[2]) && (rankList[4] == rankList[5])) ||
           ((rankList[2] == rankList[3]) && (rankList[4] == rankList[5]))
end


# Determines if the hand is a pair.
function isPair(self::Hand)
    rankList = getRankList(self)

    for i = 1:4
        (rankList[i] == rankList[i + 1]) && return true
    end

    return false
end


# Determines the tie breaking card of the hand depending on its handType.
function getTieBreakerCard(self::Hand, pass::Int)

    if (self.handType == 10) ; return self.sorted[4]

    elseif (self.handType == 9)
        return (self.sorted[4].rank == 14 && self.sorted[0].rank == 2) ?
                self.sorted[3] : self.sorted[4]

    elseif (self.handType == 8) ; return self.sorted[2]
    elseif (self.handType == 7) ; return self.sorted[2]
    elseif (self.handType == 6) ; return self.sorted[4]
    elseif (self.handType == 5) ; return self.sorted[4]
    elseif (self.handType == 4) ; return self.sorted[2]

    elseif (self.handType == 3)
        pairList = Card[]

        for i = 2:5
            currentCard = self.sorted[i]
            previousCard = self.sorted[i - 1]

            if (currentCard.rank - previousCard.rank == 0)
                push!(pairList, currentCard)
            end
        end

        max = pairList[0]
        min = pairList[0]

        for card in pairList
            if (card.rank > max.rank) ; max = card ; end
            if (card.rank < min.rank) ; min = card ; end
        end

        if     (pass == 0) ; return max
        elseif (pass == 1) ; return min
        else               ; return getKicker()
        end

    elseif (self.handType == 2)
        ret::Union{Card, Nothing} = nothing

        for i = 2:5
            currentCard = self.sorted[i]
            previousCard = self.sorted[i - 1]

            if (currentCard.rank - previousCard.rank) == 0
                ret = currentCard
                break
            end
        end

        if     (pass == 0) ; return ret
        elseif (pass == 1) ; return getKicker(self)
        end
    end

    return self.sorted[4]
end

# Helper method to get a sorted list of ranks in the hand.
function getRankList(self::Hand)
    rankList = Int[]

    for card in self.cards
        push!(rankList, card.rank)
    end

    return sort(rankList)
end


# Gets the kicker card for pairs and two pairs.
function getKicker(self::Hand)
    nonPairList = Card[]

    for i = 2:5
        currentCard = self.sorted[i]
        previousCard = self.sorted[i - 1]

        if (compare(currentCard, previousCard) != 0)
            push!(nonPairList, currentCard)
        end
    end

    max = nonPairList[1]

    for card in nonPairList
        if (compare(card, max) > 0) ; max = card ; end
    end

    return max;
end


# Returns a sorted version of a provided hand.
function sortHand(cardList::Vector{Card})
    for j in 1:4, i in 1:4
        if (compare(cardList[i], cardList[i + 1]) > 0)
            temp            = cardList[i + 1]
            cardList[i + 1] = cardList[i]
            cardList[i]     = temp
        end
    end

    return cardList
end

File: go/Hand.go
package main
import ("sort")

// Represents a hand of playing cards.
// This class defines a hand object that can hold a collection of card objects.
// author: Davis Guest
type Hand struct {
	cards []*Card
	sorted []*Card
	handType int
}


// Initializes a new hand with an empty list of cards, its sorted variant, and a default hand type of 0.
func initHand() *Hand {
	return &Hand {
		cards    : make([]*Card, 0),
		sorted   : make([]*Card, 0),
		handType : 0,
	}
}


// Adds a card to the hand's list of cards.
func (h *Hand) addCard(card *Card) {
	h.cards = append(h.cards, card)
}


// Returns a string representation of the hand.
func (h *Hand) toString() string {
	list := ""

	for i := 1; i <= len(h.cards); i++ {
		if (h.cards[i - 1].rank != 10) { list += " " }

		list += h.cards[i - 1].toString()

		if (i == 1 || i % 5 != 0) { list += " " }
	}

	if h.handType == 0         { return list
	} else if h.handType == 10 { list += " - Royal Straight Flush"
	} else if h.handType == 9  { list += " - Straight Flush"
	} else if h.handType == 8  { list += " - Four of a Kind"
	} else if h.handType == 7  { list += " - Full House"
	} else if h.handType == 6  { list += " - Flush"
	} else if h.handType == 5  { list += " - Straight"
	} else if h.handType == 4  { list += " - Three of a Kind"
	} else if h.handType == 3  { list += " - Two Pair"
	} else if h.handType == 2  { list += " - Pair"
	} else                     { list += " - High Card" }

	return list
}


// Compares this hand with another hand based on their ranks.
// Used to sort each hand type and tiebreakers accordingly.
func (h *Hand) compare(other *Hand) int {
	h.assessHand()
	other.assessHand()

	typeComparison := h.handType - other.handType

	return h.compareHelper(other, typeComparison, 0)
}


// Recursive helper method for the compare method.
func (h *Hand) compareHelper(other *Hand, diff int, pass int) int {
	if (diff != 0) { return diff }

	thisBreaker := h.getTieBreakerCard(pass)
	otherBreaker := other.getTieBreakerCard(pass)

	rankDiff := thisBreaker.rank - otherBreaker.rank
	suitDiff := thisBreaker.suit - otherBreaker.suit

	if (pass < 2 && (h.handType == 2 || h.handType == 3)) {
		if (rankDiff == 0) {
			return h.compareHelper(other, rankDiff, pass + 1)
		}
		return rankDiff
	}

	if (rankDiff == 0) { return suitDiff }
	return rankDiff
}


// Analyzes the current collection of cards in the hand and determines its hand type.
// The sorted instance variable is updated to contain the cards sorted by rank.
// This method sets the handType instance variable.
func (h *Hand) assessHand() {
	h.sortHand()

	if (h.isRoyalStraightFlush())   { h.handType = 10
	} else if (h.isStraightFlush()) { h.handType = 9
	} else if (h.isFourOfAKind())   { h.handType = 8
	} else if (h.isFullHouse())     { h.handType = 7
	} else if (h.isFlush())         { h.handType = 6
	} else if (h.isStraight())      { h.handType = 5
	} else if (h.isThreeOfAKind())  { h.handType = 4
	} else if (h.isTwoPair())       { h.handType = 3
	} else if (h.isPair())          { h.handType = 2
	} else                          { h.handType = 1 }
}


// Determines if the hand is a royal straight flush.
func (h *Hand) isRoyalStraightFlush() bool {
	rankList := h.getRankList()

	return h.isStraightFlush() &&
		   rankList[0] == 10   &&
		   rankList[4] == 14
}


// Determines if the hand is a straight flush.
func (h *Hand) isStraightFlush() bool {
	return h.isStraight() && h.isFlush()
}


// Determines if the hand is a four of a kind.
func (h *Hand) isFourOfAKind() bool {
	rankList := h.getRankList()

	return (rankList[0] == rankList[3]) ||
		   (rankList[1] == rankList[4])
}


// Determines if the hand is a full house.
func (h *Hand) isFullHouse() bool {
	rankList := h.getRankList()

	return (rankList[0] == rankList[1] && rankList[2] == rankList[4]) ||
		   (rankList[0] == rankList[2] && rankList[3] == rankList[4])
}


// Determines if the hand is a flush.
func (h *Hand) isFlush() bool {
	suit := h.sorted[0].suit

	for i := 0; i < 4; i++ {
		if h.sorted[i].suit != suit {
			return false
		}
	}

	return h.sorted[4].suit == suit
}


// Determines if the hand is a straight.
func (h *Hand) isStraight() bool {
	rankList := h.getRankList()

	if rankList[4] == 14 && rankList[0] == 2 {
		rankList[4] = 1
		sort.Ints(rankList)
	}

	for i := 0; i < 4; i++ {
		if rankList[i+1] != rankList[i]+1 {
			return false
		}
	}

	return true
}


// Determines if the hand is a three of a kind.
func (h *Hand) isThreeOfAKind() bool {
	rankList := h.getRankList()

	return (rankList[0] == rankList[2]) ||
		   (rankList[1] == rankList[3]) ||
		   (rankList[2] == rankList[4])
}


// Determines if the hand is a two pair.
func (h *Hand) isTwoPair() bool {
	rankList := h.getRankList()

	return (rankList[0] == rankList[1] && rankList[2] == rankList[3]) ||
		   (rankList[0] == rankList[1] && rankList[3] == rankList[4]) ||
		   (rankList[1] == rankList[2] && rankList[3] == rankList[4])
}


// Determines if the hand is a pair.
func (h *Hand) isPair() bool {
	rankList := h.getRankList()

	for i := 0; i < 4; i++ {
		if rankList[i] == rankList[i+1] {
			return true
		}
	}

	return false
}


// Determines the tie breaking card of the hand depending on its handType.
func (h *Hand) getTieBreakerCard(pass int) *Card {
	if h.handType == 10 { return h.sorted[4]

	} else if h.handType == 9 {
		if h.sorted[4].rank == 14 && h.sorted[0].rank == 2 {
			return h.sorted[3]
		}
		return h.sorted[4]

	} else if h.handType == 8 { return h.sorted[2]
	} else if h.handType == 7 { return h.sorted[2]
	} else if h.handType == 6 { return h.sorted[4]
	} else if h.handType == 5 { return h.sorted[4]
	} else if h.handType == 4 { return h.sorted[2]

	} else if h.handType == 3 {
		pairList := make([]*Card, 0)

		for i := 1; i < len(h.sorted); i++ {
			currentCard  := h.sorted[i]
			previousCard := h.sorted[i - 1]

			if currentCard.rank == previousCard.rank {
				pairList = append(pairList, currentCard)
			}
		}

		max := pairList[0]
		min := pairList[0]

		for _, card := range pairList {
			if card.rank > max.rank { max = card }
			if card.rank < min.rank { min = card }
		}

		if pass == 0        { return max
		} else if pass == 1 { return min
		} else              { return h.getKicker() }

	} else if h.handType == 2 {
		var ret *Card = nil

		for i := 1; i < len(h.sorted); i++ {
			currentCard  := h.sorted[i]
			previousCard := h.sorted[i - 1]

			if currentCard.rank == previousCard.rank {
				ret = currentCard
				break
			}
		}

		if pass == 0 { return ret
		} else if pass == 1 { return h.getKicker() }
	}

	return h.sorted[4]
}

// Helper method to get a sorted list of ranks in the hand.
func (h *Hand) getRankList() []int {
	rankList := make([]int, 0)

	for i := 0; i < len(h.cards); i++ {
		rankList = append(rankList, h.cards[i].rank)
	}

	sort.Ints(rankList)

	return rankList;
}

// Gets the kicker card for pairs and two pairs.
func (h *Hand) getKicker() *Card {
	nonPairList := make([]*Card, 0)

	for i := 1; i < len(h.sorted); i++ {
		currentCard  := h.sorted[i]
		previousCard := h.sorted[i - 1]

		if currentCard.rank != previousCard.rank {
			nonPairList = append(nonPairList, currentCard)
		}
	}

	max := nonPairList[0]

	for _, card := range nonPairList {
		if card.rank > max.rank { max = card }
	}

	return max
}


// Sets the sorted instance variable to a sorted version of a provided hand.
func (h *Hand) sortHand() {

	for i := 0; i < 5; i++ {
		h.sorted = append(h.sorted, initCard(h.cards[i].rank, h.cards[i].suit))
	}
	
	for j := 0; j < 4; j++ {
		for i := 0; i < 4; i++ {
			if h.sorted[i].compare(h.sorted[i + 1]) > 0 {
				temp := initCard(h.sorted[i + 1].rank, h.sorted[i + 1].suit)
				h.sorted[i + 1] = h.sorted[i]
				h.sorted[i] = temp
			}
		}
	}
}